lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,683 @@
+    unsigned implIdx = 0;
+    for (auto path : paths) {
+      builder.setInsertionPointAfter(blackBox.extModule);
+
+      // Create a new firrtl.module that implements the data tap.
+      auto name = StringAttr::get(
+          &getContext(),
+          llvm::formatv("{0}_impl_{1}", blackBox.extModule.getName(), implIdx++)
+              .str());
+      LLVM_DEBUG(llvm::dbgs()
+                 << "Implementing " << name << " ("
+                 << blackBox.extModule.getName() << " for " << path << ")\n");
+      auto impl = builder.create<FModuleOp>(blackBox.extModule->getLoc(), name,
+                                            blackBox.extModule.getPorts(),
+                                            blackBox.filteredModuleAnnos);
+      builder.setInsertionPointToEnd(impl.getBodyBlock());
+
+      // Connect the output ports to the appropriate tapped object.
+      for (auto port : portWiring) {
+        LLVM_DEBUG(llvm::dbgs() << "- Wiring up port " << port.portNum << "\n");
+        // Determine the shortest hierarchical prefix from this black box
+        // instance to the tapped object.
+        Optional<InstancePath> shortestPrefix;
+        for (auto prefix : port.prefices) {
+          auto relative = stripCommonPrefix(prefix, path);
+          if (!shortestPrefix.hasValue() ||
+              relative.size() < shortestPrefix->size())
+            shortestPrefix = relative;
+        }
+        if (!shortestPrefix.hasValue()) {
+          LLVM_DEBUG(llvm::dbgs() << "  - Has no prefix, skipping\n");
+          continue;
+        }
+        LLVM_DEBUG(llvm::dbgs()
+                   << "  - Shortest prefix " << *shortestPrefix << "\n");
+
+        // Concatenate the prefix into a proper full hierarchical name.
+        SmallString<128> hname;
+        for (auto inst : shortestPrefix.getValue()) {
+          if (!hname.empty())
+            hname += '.';
+          hname += inst.name();
+        }
+        if (!hname.empty())
+          hname += '.';
+        hname += port.suffix;
+        LLVM_DEBUG(llvm::dbgs() << "  - Connecting as " << hname << "\n");
+
+        // Add a verbatim op that assigns this module port.
+        auto portName = getModulePortName(blackBox.extModule, port.portNum);
+        auto arg = impl.getArgument(port.portNum);
+        auto argType = arg.getType().cast<FIRRTLType>();
+        auto width = argType.getBitWidthOrSentinel();
+        if (width < 0) {
+          blackBox.extModule->emitError(llvm::formatv(
+              "data/mem tap \"{0}\" has unsupported type {1} (module port {2})",
+              hname, argType, portName));
+          signalPassFailure();
+          continue;
+        }
+
+        // Create the `connect(port, cast(verbatim("foo.bar")))` chain.
+        auto intType = builder.getIntegerType(width);
+        auto hnameVerbatim = builder.create<sv::VerbatimExprOp>(
+            blackBox.extModule->getLoc(), intType, hname);
+        auto hnameCast = builder.create<StdIntCastOp>(
```

Comment:
plz don't use StdIntCast outside LowerToHW.  i'd like to eventually remove it.

Reply:
Done. Replaced with a more appropriate `firrtl.verbatim.expr` as you suggested (separate PR).